plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ServerEvaluationStrategy.java

```diff
@@ -0,0 +1,197 @@
+import org.eclipse.che.api.machine.server.model.impl.ServerConfImpl;
+import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
+import org.eclipse.che.api.machine.server.model.impl.ServerPropertiesImpl;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
+import org.eclipse.che.plugin.docker.client.json.PortBinding;
+
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+
+/**
+ * Represents a strategy for resolving Servers associated with workspace containers.
+ * Used to extract relevant information from e.g. {@link ContainerInfo} into a
+ * {@link ServerImpl} object.
+ *
+ * @author Angel Misevski <amisevsk@redhat.com>
+ * @see ServerEvaluationStrategyProvider
+ */
+public abstract class ServerEvaluationStrategy {
+
+    protected static final String SERVER_CONF_LABEL_REF_KEY      = "che:server:%s:ref";
+    protected static final String SERVER_CONF_LABEL_PROTOCOL_KEY = "che:server:%s:protocol";
+    protected static final String SERVER_CONF_LABEL_PATH_KEY     = "che:server:%s:path";
+
+    /**
+     * Used to store the address set by property {@code che.docker.ip}, if applicable.
+     */
+    protected String internalAddressProperty;
+
+    /**
+     * Used to store the address set by property {@code che.docker.ip.external}. if applicable.
+     */
+    protected String externalAddressProperty;
+
+    /**
+     * Gets a map of all <strong>internal</strong> addresses exposed by the container in the form of
+     * {@code "<address>:<port>"}
+     *
+     * @param containerInfo the ContainerInfo object that describes the container.
+     * @param internalAddress address passed into {@code getServers}; used as fallback if address cannot be
+     *        retrieved from containerInfo.
+     * @return a Map of port protocol (e.g. "4401/tcp") to address (e.g. "172.17.0.1:32317")
+     */
+    protected abstract Map<String, String> getInternalAddressesAndPorts(ContainerInfo containerInfo,
+                                                                        String internalAddress);
+
+    /**
+     * Gets a map of all <strong>external</strong> addresses exposed by the container in the form of
+     * {@code "<address>:<port>"}
+     *
+     * @param containerInfo the ContainerInfo object that describes the container.
+     * @param internalAddress address passed into {@code getServers}; used as fallback if address cannot be
+     *        retrieved from containerInfo.
+     * @return a Map of port protocol (e.g. "4401/tcp") to address (e.g. "localhost:32317")
+     */
+    protected abstract Map<String, String> getExternalAddressesAndPorts(ContainerInfo containerInfo,
+                                                                        String internalAddress);
+
+    @Inject
+    protected ServerEvaluationStrategy(@Nullable @Named("che.docker.ip")          String internalAddressProperty,
```

Comment:
We put constructor before other methods in Che source code. Please follow that to not distract other devs.